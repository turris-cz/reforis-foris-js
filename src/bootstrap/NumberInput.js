/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { faMinus, faPlus } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import PropTypes from "prop-types";

import Input from "./Input";
import { useConditionalTimeout } from "../utils/hooks";
import "./NumberInput.css";

NumberInput.propTypes = {
    /** Field label. */
    label: PropTypes.string.isRequired,
    /** Error message. */
    error: PropTypes.string,
    /** Help text message. */
    helpText: PropTypes.string,
    /** Number value. */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Function called when value changes. */
    onChange: PropTypes.func.isRequired,
    /** Additional description displayed to the right of input value. */
    inlineText: PropTypes.string,
};

NumberInput.defaultProps = {
    value: 0,
};

function NumberInput({ onChange, inlineText, value, ...props }) {
    function updateValue(initialValue, difference) {
        onChange({ target: { value: initialValue + difference } });
    }

    const enableIncrease = useConditionalTimeout(
        { callback: updateValue },
        value,
        1
    );
    const enableDecrease = useConditionalTimeout(
        { callback: updateValue },
        value,
        -1
    );

    return (
        <Input type="number" onChange={onChange} value={value} {...props}>
            {inlineText && (
                <span className="input-group-text">{inlineText}</span>
            )}
            <button
                type="button"
                className="btn btn-outline-secondary"
                onMouseDown={() => enableIncrease(true)}
                onMouseUp={() => enableIncrease(false)}
                aria-label="Increase"
            >
                <FontAwesomeIcon icon={faPlus} />
            </button>
            <button
                type="button"
                className="btn btn-outline-secondary"
                onMouseDown={() => enableDecrease(true)}
                onMouseUp={() => enableDecrease(false)}
                aria-label="Decrease"
            >
                <FontAwesomeIcon icon={faMinus} />
            </button>
        </Input>
    );
}

export default NumberInput;
