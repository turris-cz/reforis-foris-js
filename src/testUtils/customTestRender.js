/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from "react";
import { UIDReset } from "react-uid";
import { StaticRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";

import { AlertContextMock } from "./alertContextMock";
import { CustomizationContextMock } from "./customizationContextMock";

Wrapper.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

function Wrapper({ children }) {
    return (
        <CustomizationContextMock>
            <AlertContextMock>
                <StaticRouter>
                    <UIDReset>{children}</UIDReset>
                </StaticRouter>
            </AlertContextMock>
        </CustomizationContextMock>
    );
}

const customTestRender = (ui, options) =>
    render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customTestRender as render };
