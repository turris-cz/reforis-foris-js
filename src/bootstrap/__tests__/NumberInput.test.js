/*
 * Copyright (C) 2019-2025 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { render, fireEvent, getByLabelText, waitFor } from "customTestRender";

import NumberInput from "../NumberInput";

describe("<NumberInput/>", () => {
    const onChangeMock = jest.fn();
    let componentContainer;

    beforeEach(() => {
        const { container } = render(
            <NumberInput
                label="Test label"
                helpText="Some help text"
                value={1}
                onChange={onChangeMock}
            />
        );
        componentContainer = container;
    });

    it("Render number input", () => {
        expect(componentContainer.firstChild).toMatchSnapshot();
    });

    it("Increase number with button", async () => {
        const increaseButton = getByLabelText(componentContainer, /Increase/);
        fireEvent.mouseDown(increaseButton);
        await waitFor(() =>
            expect(onChangeMock).toHaveBeenCalledWith({ target: { value: 2 } })
        );
    });

    it("Decrease number with button", async () => {
        const decreaseButton = getByLabelText(componentContainer, /Decrease/);
        fireEvent.mouseDown(decreaseButton);
        await waitFor(() =>
            expect(onChangeMock).toHaveBeenCalledWith({ target: { value: 0 } })
        );
    });
});
