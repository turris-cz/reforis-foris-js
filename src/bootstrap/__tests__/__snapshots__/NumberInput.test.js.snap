// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput/> Render number input 1`] = `
<div
  class="mb-3"
>
  <label
    class="form-label"
    for="1"
  >
    Test label
  </label>
  <div
    class="input-group"
  >
    <input
      class="form-control"
      id="1"
      type="number"
      value="1"
    />
    <button
      aria-label="Increase"
      class="btn btn-outline-secondary"
      type="button"
    >
      <i
        class="fas fa-plus"
      />
    </button>
    <button
      aria-label="Decrease"
      class="btn btn-outline-secondary"
      type="button"
    >
      <i
        class="fas fa-minus"
      />
    </button>
  </div>
  <div
    class="form-text"
  >
    <small>
      Some help text
    </small>
  </div>
</div>
`;
